File: classified/packages/plugin-inference/src/claude-proxy.ts
import { IAgentRuntime, ModelType, logger } from '@elizaos/core';
import { InferenceProvider } from './index';

// Claude API Types
interface ClaudeTool {
  name: string;
  description?: string;
  input_schema: any;
}

type ClaudeContent =
  | string
  | Array<{
      type: 'text' | 'image' | 'tool_use' | 'tool_result';
      text?: string;
      source?: {
        type: 'base64';
        media_type: string;
        data: string;
      };
      id?: string;
      name?: string;
      input?: any;
      tool_use_id?: string;
      content?: any;
    }>;

interface ClaudeMessage {
  role: 'user' | 'assistant';
  content: ClaudeContent;
}

interface ClaudeMessagesRequest {
  model: string;
  messages: ClaudeMessage[];
  system?: string;
  max_tokens: number;
  stop_sequences?: string[];
  stream?: boolean;
  temperature?: number;
  top_p?: number;
  top_k?: number;
  tools?: ClaudeTool[];
  tool_choice?: { type: 'auto' | 'any' | 'tool'; name?: string };
}

interface ClaudeResponse {
  id: string;
  type: 'message';
  role: 'assistant';
  model: string;
  content: Array<{
    type: 'text' | 'tool_use';
    text?: string;
    id?: string;
    name?: string;
    input?: any;
  }>;
  stop_reason: string;
  usage: {
    input_tokens: number;
    output_tokens: number;
  };
}

// OpenAI-style types for compatibility
interface OpenAIMessage {
  role: 'system' | 'user' | 'assistant' | 'tool';
  content:
    | string
    | Array<{ type: 'text' | 'image_url'; text?: string; image_url?: { url: string } }>;
  tool_calls?: Array<{
    id: string;
    type: 'function';
    function: {
      name: string;
      arguments: string;
    };
  }>;
  tool_call_id?: string;
}

/**
 * Claude-to-Provider Proxy
 * Converts Claude API requests to the appropriate provider format
 */
export class ClaudeProxy {
  constructor(private runtime: IAgentRuntime) {}

  /**
   * Process a Claude API request and route it to the appropriate provider
   */
  async processRequest(claudeRequest: ClaudeMessagesRequest): Promise<ClaudeResponse> {
    // Determine which provider to use
    const provider = await this.selectProvider();

    logger.info(`[CLAUDE_PROXY] Routing request to provider: ${provider}`);

    switch (provider) {
      case InferenceProvider.ANTHROPIC:
        // Native Anthropic - make direct API request
        return this.makeAnthropicProxyRequest(claudeRequest);

      case InferenceProvider.OPENAI:
        return this.handleOpenAIRequest(claudeRequest);

      case InferenceProvider.GROQ:
        return this.handleGroqRequest(claudeRequest);

      case InferenceProvider.OLLAMA:
        return this.handleOllamaRequest(claudeRequest);

      default:
        throw new Error(`Unsupported provider: ${provider}`);
    }
  }

  /**
   * Make a direct proxy request to Anthropic's API
   */
  private async makeAnthropicProxyRequest(
    claudeRequest: ClaudeMessagesRequest
  ): Promise<ClaudeResponse> {
    const apiKey = this.runtime.getSetting('ANTHROPIC_API_KEY');
    if (!apiKey) {
      throw new Error('ANTHROPIC_API_KEY not configured');
    }

    const baseUrl = this.runtime.getSetting('ANTHROPIC_API_URL') || 'https://api.anthropic.com';

    try {
      // Make direct request to Anthropic API
      const response = await fetch(`${baseUrl}/v1/messages`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'x-api-key': apiKey,
          'anthropic-version': '2023-06-01',
        },
        body: JSON.stringify(claudeRequest),
      });

      if (!response.ok) {
        const errorBody = await response.text();
        throw new Error(`Anthropic API error: ${response.status} - ${errorBody}`);
      }

      const data = (await response.json()) as ClaudeResponse;
      return data;
    } catch (error) {
      logger.error('[CLAUDE_PROXY] Error making Anthropic proxy request:', error);
      throw error;
    }
  }

  /**
   * Select the best available provider
   */
  private async selectProvider(): Promise<InferenceProvider> {
    // Check available providers in order of preference
    const preferences = [
      InferenceProvider.ANTHROPIC,
      InferenceProvider.OPENAI,
      InferenceProvider.GROQ,
      InferenceProvider.OLLAMA,
    ];

    for (const provider of preferences) {
      if (await this.isProviderAvailable(provider)) {
        return provider;
      }
    }

    throw new Error('No inference providers available');
  }

  /**
   * Check if a provider is available
   */
  private async isProviderAvailable(provider: InferenceProvider): Promise<boolean> {
    switch (provider) {
      case InferenceProvider.ANTHROPIC:
        return !!this.runtime.getSetting('ANTHROPIC_API_KEY');
      case InferenceProvider.OPENAI:
        return !!this.runtime.getSetting('OPENAI_API_KEY');
      case InferenceProvider.GROQ:
        return !!this.runtime.getSetting('GROQ_API_KEY');
      case InferenceProvider.OLLAMA:
        // Ollama is always available if installed
        return true;
      default:
        return false;
    }
  }

  /**
   * Handle OpenAI requests by converting format
   */
  private async handleOpenAIRequest(claudeRequest: ClaudeMessagesRequest): Promise<ClaudeResponse> {
    const openAIMessages = this.convertClaudeToOpenAI(claudeRequest);

    // Create a formatted prompt for the runtime
    const prompt = this.formatOpenAIPrompt(openAIMessages, claudeRequest.system);

    const response = await this.runtime.useModel(ModelType.TEXT_LARGE, {
      runtime: this.runtime,
      prompt,
      modelType: ModelType.TEXT_LARGE,
      maxTokens: claudeRequest.max_tokens,
      temperature: claudeRequest.temperature,
      stopSequences: claudeRequest.stop_sequences,
    });

    return this.formatClaudeResponse(response, claudeRequest.model);
  }

  /**
   * Handle Groq requests
   */
  private async handleGroqRequest(claudeRequest: ClaudeMessagesRequest): Promise<ClaudeResponse> {
    // Groq uses OpenAI-compatible format
    return this.handleOpenAIRequest(claudeRequest);
  }

  /**
   * Handle Ollama requests
   */
  private async handleOllamaRequest(claudeRequest: ClaudeMessagesRequest): Promise<ClaudeResponse> {
    // Ollama also uses OpenAI-compatible format
    const prompt = this.convertClaudeToPrompt(claudeRequest);

    const response = await this.runtime.useModel(ModelType.TEXT_LARGE, {
      runtime: this.runtime,
      prompt,
      modelType: ModelType.TEXT_LARGE,
      maxTokens: claudeRequest.max_tokens,
      temperature: claudeRequest.temperature,
      stopSequences: claudeRequest.stop_sequences,
    });

    return this.formatClaudeResponse(response, claudeRequest.model);
  }

  /**
   * Convert Claude messages to OpenAI format
   */
  private convertClaudeToOpenAI(claudeRequest: ClaudeMessagesRequest): OpenAIMessage[] {
    const openAIMessages: OpenAIMessage[] = [];

    if (claudeRequest.system) {
      openAIMessages.push({ role: 'system', content: claudeRequest.system });
    }

    for (const message of claudeRequest.messages) {
      if (message.role === 'user') {
        if (Array.isArray(message.content)) {
          const content = message.content
            .filter((c) => c.type === 'text')
            .map((c) => c.text)
            .join('\n');
          openAIMessages.push({ role: 'user', content });
        } else {
          openAIMessages.push({ role: 'user', content: message.content });
        }
      } else if (message.role === 'assistant') {
        if (Array.isArray(message.content)) {
          const textContent = message.content
            .filter((c) => c.type === 'text')
            .map((c) => c.text)
            .join('\n');
          openAIMessages.push({ role: 'assistant', content: textContent || '' });
        } else {
          openAIMessages.push({ role: 'assistant', content: message.content });
        }
      }
    }

    return openAIMessages;
  }

  /**
   * Convert Claude messages to a simple prompt string
   */
  private convertClaudeToPrompt(claudeRequest: ClaudeMessagesRequest): string {
    let prompt = '';

    if (claudeRequest.system) {
      prompt += `System: ${claudeRequest.system}\n\n`;
    }

    for (const message of claudeRequest.messages) {
      const role = message.role === 'user' ? 'User' : 'Assistant';

      if (Array.isArray(message.content)) {
        const textContent = message.content
          .filter((c) => c.type === 'text')
          .map((c) => c.text)
          .join('\n');
        prompt += `${role}: ${textContent}\n\n`;
      } else {
        prompt += `${role}: ${message.content}\n\n`;
      }
    }

    return prompt.trim();
  }

  /**
   * Format OpenAI messages as a prompt
   */
  private formatOpenAIPrompt(messages: OpenAIMessage[], system?: string): string {
    let prompt = '';

    if (system) {
      prompt += `${system}\n\n`;
    }

    for (const message of messages) {
      if (message.role === 'system') {
        continue; // Already handled above
      }

      const role = message.role === 'user' ? 'User' : 'Assistant';
      const content =
        typeof message.content === 'string'
          ? message.content
          : message.content.map((c) => c.text).join('\n');

      prompt += `${role}: ${content}\n\n`;
    }

    return prompt.trim();
  }

  /**
   * Format response in Claude format
   */
  private formatClaudeResponse(text: string, model: string): ClaudeResponse {
    return {
      id: `msg_${Math.random().toString(36).substr(2, 9)}`,
      type: 'message',
      role: 'assistant',
      model,
      content: [
        {
          type: 'text',
          text,
        },
      ],
      stop_reason: 'end_turn',
      usage: {
        input_tokens: 0, // Would need token counting
        output_tokens: 0, // Would need token counting
      },
    };
  }
}

/**
 * Create a Claude API handler for the inference plugin
 */
export function createClaudeHandler(runtime: IAgentRuntime) {
  const proxy = new ClaudeProxy(runtime);

  return async (request: ClaudeMessagesRequest): Promise<ClaudeResponse> => {
    try {
      return await proxy.processRequest(request);
    } catch (error) {
      logger.error('[CLAUDE_PROXY] Error processing request:', error);
      throw error;
    }
  };
}

File: cpp_merge/classified/packages/plugin-inference/src/claude-proxy.cpp
#include "/home/runner/work/elizaos-cpp/elizaos-cpp/classified/packages/plugin-inference/src/claude-proxy.h"

ClaudeProxy::ClaudeProxy(std::shared_ptr<IAgentRuntime> runtime_) : runtime(runtime_)  {
}

std::shared_ptr<Promise<std::shared_ptr<ClaudeResponse>>> ClaudeProxy::processRequest(std::shared_ptr<ClaudeMessagesRequest> claudeRequest)
{
    auto provider = std::async([=]() { this->selectProvider(); });
    logger->info(std:("[CLAUDE_PROXY] Routing request to provider: ") + provider + string_empty);
    switch (provider)
    {
    case InferenceProvider::ANTHROPIC:
        return this->makeAnthropicProxyRequest(claudeRequest);
    case InferenceProvider::OPENAI:
        return this->handleOpenAIRequest(claudeRequest);
    case InferenceProvider::GROQ:
        return this->handleGroqRequest(claudeRequest);
    case InferenceProvider::OLLAMA:
        return this->handleOllamaRequest(claudeRequest);
    default:
        throw any(std::make_shared<Error>(std:("Unsupported provider: ") + provider + string_empty));
    }
}

std::shared_ptr<Promise<std::shared_ptr<ClaudeResponse>>> ClaudeProxy::makeAnthropicProxyRequest(std::shared_ptr<ClaudeMessagesRequest> claudeRequest)
{
    auto apiKey = this->runtime->getSetting(std:("ANTHROPIC_API_KEY"));
    if (!apiKey) {
        throw any(std::make_shared<Error>(std:("ANTHROPIC_API_KEY not configured")));
    }
    auto baseUrl = OR((this->runtime->getSetting(std:("ANTHROPIC_API_URL"))), (std:("https://api.anthropic.com")));
    try
    {
        auto response = std::async([=]() { fetch(string_empty + baseUrl + std:("/v1/messages"), object{
            object::pair{std:("method"), std:("POST")}, 
            object::pair{std:("headers"), object{
                object::pair{std:("Content-Type"), std:("application/json")}, 
                object::pair{std:("x-api-key"), apiKey}, 
                object::pair{std:("anthropic-version"), std:("2023-06-01")}
            }}, 
            object::pair{std:("body"), JSON->stringify(claudeRequest)}
        }); });
        if (!response->ok) {
            auto errorBody = std::async([=]() { response->text(); });
            throw any(std::make_shared<Error>(std:("Anthropic API error: ") + response->status + std:(" - ") + errorBody + string_empty));
        }
        auto data = as<std::shared_ptr<ClaudeResponse>>((std::async([=]() { response->json(); })));
        return data;
    }
    catch (const any& error)
    {
        logger->error(std:("[CLAUDE_PROXY] Error making Anthropic proxy request:"), error);
        throw any(error);
    }
}

std::shared_ptr<Promise<InferenceProvider>> ClaudeProxy::selectProvider()
{
    auto preferences = array<InferenceProvider>{ InferenceProvider::ANTHROPIC, InferenceProvider::OPENAI, InferenceProvider::GROQ, InferenceProvider::OLLAMA };
    for (auto& provider : preferences)
    {
        if (std::async([=]() { this->isProviderAvailable(provider); })) {
            return provider;
        }
    }
    throw any(std::make_shared<Error>(std:("No inference providers available")));
}

std::shared_ptr<Promise<boolean>> ClaudeProxy::isProviderAvailable(InferenceProvider provider)
{
    switch (provider)
    {
    case InferenceProvider::ANTHROPIC:
        return !!this->runtime->getSetting(std:("ANTHROPIC_API_KEY"));
    case InferenceProvider::OPENAI:
        return !!this->runtime->getSetting(std:("OPENAI_API_KEY"));
    case InferenceProvider::GROQ:
        return !!this->runtime->getSetting(std:("GROQ_API_KEY"));
    case InferenceProvider::OLLAMA:
        return true;
    default:
        return false;
    }
}

std::shared_ptr<Promise<std::shared_ptr<ClaudeResponse>>> ClaudeProxy::handleOpenAIRequest(std::shared_ptr<ClaudeMessagesRequest> claudeRequest)
{
    auto openAIMessages = this->convertClaudeToOpenAI(claudeRequest);
    auto prompt = this->formatOpenAIPrompt(openAIMessages, claudeRequest->system);
    auto response = std::async([=]() { this->runtime->useModel(ModelType->TEXT_LARGE, object{
        object::pair{std:("runtime"), this->runtime}, 
        object::pair{std:("prompt"), std:("prompt")}, 
        object::pair{std:("modelType"), ModelType->TEXT_LARGE}, 
        object::pair{std:("maxTokens"), claudeRequest->max_tokens}, 
        object::pair{std:("temperature"), claudeRequest->temperature}, 
        object::pair{std:("stopSequences"), claudeRequest->stop_sequences}
    }); });
    return this->formatClaudeResponse(response, claudeRequest->model);
}

std::shared_ptr<Promise<std::shared_ptr<ClaudeResponse>>> ClaudeProxy::handleGroqRequest(std::shared_ptr<ClaudeMessagesRequest> claudeRequest)
{
    return this->handleOpenAIRequest(claudeRequest);
}

std::shared_ptr<Promise<std::shared_ptr<ClaudeResponse>>> ClaudeProxy::handleOllamaRequest(std::shared_ptr<ClaudeMessagesRequest> claudeRequest)
{
    auto prompt = this->convertClaudeToPrompt(claudeRequest);
    auto response = std::async([=]() { this->runtime->useModel(ModelType->TEXT_LARGE, object{
        object::pair{std:("runtime"), this->runtime}, 
        object::pair{std:("prompt"), std:("prompt")}, 
        object::pair{std:("modelType"), ModelType->TEXT_LARGE}, 
        object::pair{std:("maxTokens"), claudeRequest->max_tokens}, 
        object::pair{std:("temperature"), claudeRequest->temperature}, 
        object::pair{std:("stopSequences"), claudeRequest->stop_sequences}
    }); });
    return this->formatClaudeResponse(response, claudeRequest->model);
}

array<std::shared_ptr<OpenAIMessage>> ClaudeProxy::convertClaudeToOpenAI(std::shared_ptr<ClaudeMessagesRequest> claudeRequest)
{
    auto openAIMessages = array<std::shared_ptr<OpenAIMessage>>();
    if (claudeRequest->system) {
        openAIMessages->push(object{
            object::pair{std:("role"), std:("system")}, 
            object::pair{std:("content"), claudeRequest->system}
        });
    }
    for (auto& message : claudeRequest->messages)
    {
        if (message->role == std:("user")) {
            if (Array->isArray(message->content)) {
                auto content = message->content->filter([=](auto c) mutable
                {
                    return c["type"] == std:("text");
                }
                )->map([=](auto c) mutable
                {
                    return c["text"];
                }
                )->join(std:("\
"));
                openAIMessages->push(object{
                    object::pair{std:("role"), std:("user")}, 
                    object::pair{std:("content"), std:("content")}
                });
            } else {
                openAIMessages->push(object{
                    object::pair{std:("role"), std:("user")}, 
                    object::pair{std:("content"), message->content}
                });
            }
        } else if (message->role == std:("assistant")) {
            if (Array->isArray(message->content)) {
                auto textContent = message->content->filter([=](auto c) mutable
                {
                    return c["type"] == std:("text");
                }
                )->map([=](auto c) mutable
                {
                    return c["text"];
                }
                )->join(std:("\
"));
                openAIMessages->push(object{
                    object::pair{std:("role"), std:("assistant")}, 
                    object::pair{std:("content"), OR((textContent), (string_empty))}
                });
            } else {
                openAIMessages->push(object{
                    object::pair{std:("role"), std:("assistant")}, 
                    object::pair{std:("content"), message->content}
                });
            }
        }
    }
    return openAIMessages;
}

string ClaudeProxy::convertClaudeToPrompt(std::shared_ptr<ClaudeMessagesRequest> claudeRequest)
{
    auto prompt = string_empty;
    if (claudeRequest->system) {
        prompt += std:("System: ") + claudeRequest->system + std:("\
\
");
    }
    for (auto& message : claudeRequest->messages)
    {
        auto role = (message->role == std:("user")) ? std:("User") : std:("Assistant");
        if (Array->isArray(message->content)) {
            auto textContent = message->content->filter([=](auto c) mutable
            {
                return c["type"] == std:("text");
            }
            )->map([=](auto c) mutable
            {
                return c["text"];
            }
            )->join(std:("\
"));
            prompt += string_empty + role + std:(": ") + textContent + std:("\
\
");
        } else {
            prompt += string_empty + role + std:(": ") + message->content + std:("\
\
");
        }
    }
    return prompt->trim();
}

string ClaudeProxy::formatOpenAIPrompt(array<std::shared_ptr<OpenAIMessage>> messages, string system)
{
    auto prompt = string_empty;
    if (system) {
        prompt += string_empty + system + std:("\
\
");
    }
    for (auto& message : messages)
    {
        if (message->role == std:("system")) {
            continue;
        }
        auto role = (message->role == std:("user")) ? std:("User") : std:("Assistant");
        auto content = (type_of(message->content) == std:("string")) ? message->content : message->content->map([=](auto c) mutable
        {
            return c["text"];
        }
        )->join(std:("\
"));
        prompt += string_empty + role + std:(": ") + content + std:("\
\
");
    }
    return prompt->trim();
}

std::shared_ptr<ClaudeResponse> ClaudeProxy::formatClaudeResponse(string text, string model)
{
    return object{
        object::pair{std:("id"), std:("msg_") + Math->random()->toString(36)->substr(2, 9) + string_empty}, 
        object::pair{std:("type"), std:("message")}, 
        object::pair{std:("role"), std:("assistant")}, 
        object::pair{std:("model"), std:("model")}, 
        object::pair{std:("content"), array<object>{ object{
            object::pair{std:("type"), std:("text")}, 
            object::pair{std:("text"), std:("text")}
        } }}, 
        object::pair{std:("stop_reason"), std:("end_turn")}, 
        object::pair{std:("usage"), object{
            object::pair{std:("input_tokens"), 0}, 
            object::pair{std:("output_tokens"), 0}
        }}
    };
}

any createClaudeHandler(std::shared_ptr<IAgentRuntime> runtime)
{
    shared proxy = std::make_shared<ClaudeProxy>(runtime);
    return [=](auto request) mutable
    {
        try
        {
            return std::async([=]() { proxy->processRequest(request); });
        }
        catch (const any& error)
        {
            logger->error(std:("[CLAUDE_PROXY] Error processing request:"), error);
            throw any(error);
        }
    };
};


